src/Microsoft.SymbolStore/KeyGenerators/ELFFileKeyGenerator.cs

```diff
@@ -113,57 +112,59 @@ public override IEnumerable<SymbolStoreKey> GetKeys(KeyTypeFlags flags)
+                        yield return BuildKey(path, SymbolPrefix, buildId, "_.debug");
+                    }
+                    else
+                    {
+                        bool clrSpecialFile = s_coreClrSpecialFiles.Contains(fileName);
+                        yield return BuildKey(path, IdentityPrefix, buildId, clrSpecialFile);
                     }
-                    yield return BuildKey(symbolFileName, SymbolPrefix, buildId, "_.debug");
                 }
-                if ((flags & (KeyTypeFlags.ClrKeys | KeyTypeFlags.DacDbiKeys)) != 0)
+                if (!symbolFile)
                 {
-                    /// Creates all the special CLR keys if the path is the coreclr module for this platform
-                    if (fileName == CoreClrFileName)
+                    // This is a workaround for 5.0 where the ELF file type of dotnet isn't Executable but 
+                    // Shared. It doesn't work for self-contained apps (apphost renamed to host program).
+                    if ((flags & KeyTypeFlags.HostKeys) != 0 && fileName == "dotnet")
+                    {
+                        yield return BuildKey(path, IdentityPrefix, buildId, clrSpecialFile: false);
+                    }
+                    if ((flags & KeyTypeFlags.RuntimeKeys) != 0 && fileName == CoreClrFileName)
+                    {
+                        yield return BuildKey(path, IdentityPrefix, buildId);
+                    }
+                    if ((flags & KeyTypeFlags.SymbolKey) != 0)
                     {
-                        foreach (string specialFileName in (flags & KeyTypeFlags.ClrKeys) != 0 ? s_coreClrSpecialFiles : s_dacdbiSpecialFiles)
+                        if (string.IsNullOrEmpty(symbolFileName))
                         {
-                            yield return BuildKey(specialFileName, CoreClrPrefix, buildId);
+                            symbolFileName = path + s_symbolFileExtensions[0];
+                        }
+                        yield return BuildKey(symbolFileName, SymbolPrefix, buildId, "_.debug");
+                    }
+                    if ((flags & (KeyTypeFlags.ClrKeys | KeyTypeFlags.DacDbiKeys)) != 0)
+                    {
+                        // Creates all the special CLR keys if the path is the coreclr module for this platform
+                        if (fileName == CoreClrFileName)
+                        {
+                            foreach (string specialFileName in (flags & KeyTypeFlags.ClrKeys) != 0 ? s_coreClrSpecialFiles : s_dacdbiSpecialFiles)
+                            {
+                                yield return BuildKey(specialFileName, CoreClrPrefix, buildId);
+                            }
                         }
                     }
                 }
             }
+            else
+            {
+                Debug.Fail($"Invalid ELF BuildId '{(buildId == null ? "<null>" : ToHexString(buildId))}' for {path}");
```

Comment:
Shouldn't we do an error log - it feels fair for this to not enumerate anything instead of failing on a debug build. Some files will likely have invalid index and we should just not generate an index for it under the current design.

Reply:
We can't log anything to the tracer because this is a static entry point used by the diagnostics repo code and not the symstore key generators.  It previous asserted if the build id parameter was null or if it wasn't 20 bytes long so this is in keeping with the previous behavior and doesn't return any keys.

Reply:
I was wrong about my that this static GetKeys function was only called by diagnostics repo.  It look like it is also called by the instance GetKeys method but after it's NormalizeBuildId call so the assert should never hit in that path.  The static method should not be called by the key generators directly or the uploader.

Reply:
Unless we explicitly make this private, you can't ensure that. What tells you symbol.exe wouldn't take a dependency on this? Why is this a necessary stati public method and not just always an instance method?

Reply:
It is a public static entry point now so it can't be made private unless we want to break things. The diag repo code (and probably others) needs a key generator entry point that takes a build id, etc. to build the key from. The KeyGenerator instance classes are went to read the build id from the ELF or MachO binary. 

In the end, if the build id is invalid, the enumerable returned will have no items so that is a detectable failure.